### DISCUSSION TITLE
Using YOLOv5 with Pillow-SIMD

### DISCUSSION THREAD

**Discussion by jkocherhans** [ultralytics/yolov5#7472] [Ideas]
Hello! I’d like to use [Pillow-SIMD](https://github.com/uploadcare/pillow-simd) (a drop-in replacement for Pillow) along with YOLOv5, but as soon as I import my model with `torch.hub.load`, `check_requirements` notices that Pillow isn’t installed and installs it automatically over the top of Pillow-SIMD. I can think of a few possible ways around this, but they all involve changing YOLOv5’s code.

1. Stop installing dependencies and print out a warning instead. Basically this would change the `install=True` default in `check_requirements` to `False`. In a production environment, it’s surprising to me that importing code would cause a package to be installed, but I definitely see how the current behavior is helpful for most folks.
2. Allow users to pass `install=False` into `torch.hub.load`, and pass that down to the call to `check_requirements`.
3. Special case `check_requirements` to check for Pillow-SIMD as a fallback for Pillow.

If any combination of those options sound like a good idea to the maintainers, I’m happy to put together a PR.

Thank you!

**Comment by glenn-jocher** [comment]
@jkocherhans you could create a PR for an environment variable that `check_requirements()` would use if found, i.e. `YOLOv5_AUTOINSTALL=False`

In any case the last time I checked PIL had already incorporated all of the speed improvements found in SIMD. Can you provide any current benchmarks with profiling results on each?

**Reply by jkocherhans** [reply]
@glenn-jocher yep! I can put together a PR to use an environment variable. That sounds more reliable than the other options I'd thought of.

I hadn't heard that Pillow had incorporated any of the SIMD changes, and I haven't run any benchmarks, but I'll take a closer look.

Thanks!

**Comment by glenn-jocher** [comment]
@jkocherhans PR #7505 merged! Let us know if you observe performance improvements with SIMD.

**Reply by jkocherhans** [reply]
Thank you for merging that change so quickly!
I did a little more exploration with Pillow vs Pillow-SIMD, and it seems like you’re right. I was seeing a pretty big performance difference with Pillow 8.4.0 vs Pillow-SIMD 8.4.0.post0, but when I updated to Pillow 9.0.1, the differences between that and Pillow-SIMD 9.0.0.post1 were minimal for what I’m doing: cropping and resizing a lot of images using BILINEAR on an AWS c5.large or c6i.large instance.
I’m still happy to be able to pin my dependencies in production now, so I’ll be using `YOLOv5_AUTOINSTALL=False` either way. Thanks again!

**Reply by glenn-jocher** [reply]
@jkocherhans oh great! Thanks for checking.
